packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -62,11 +46,211 @@ class PlaceholderSpanIndexSemanticsTag extends SemanticsTag {
   int get hashCode => Object.hash(PlaceholderSpanIndexSemanticsTag, index);
 }
 
+/// Parent data used by [RenderParagraph] and [RenderEditable] to annotate
+/// inline widgets with.
+class TextParentData extends ParentData with ContainerParentDataMixin<RenderBox> {
+  /// The offset at which to paint the child in the parent's coordinate system.
+  ///
+  /// A `null` value indicates this inline widget is not laid out. For instance,
+  /// when the inline widget has never been laid out, or the inline widget is
+  /// ellipsized away.
+  Offset? get offset => _offset;
+  Offset? _offset;
+
+  /// The [WidgetSpan] associated with this inline widget.
+  ///
+  /// This field is usually set by a [ParentDataWidget], and is typically not
+  /// null when `performLayout` is called.
+  PlaceholderSpan? span;
+
+  @override
+  void detach() {
+    span = null;
+    _offset = null;
+    super.detach();
+  }
+
+  @override
+  String toString() =>'widget: $span, ${offset == null ? "not laid out" : "offset: $offset"}';
+}
+
+/// A mixin that provides useful default behaviors for text [RenderBox]es
+/// ([RenderParagraph] and [RenderEditable] for example) with inline widget
+/// children managed by the [ContainerRenderObjectMixin] mixin.
+///
+/// This mixin assumes every child managed by the [ContainerRenderObjectMixin]
+/// mixin corresponds to a [WidgetSpan], and they are organized in logical
+/// order of the text (the order each [WidgetSpan] is encountered when the user
+/// reads the text).
+///
+/// To use this mixin in a [RenderBox] class:
+///
+///  * Call [layoutInlineWidgets] in the `performLayout` and `computeDryLayout`
+///    implementation, and during intrinsic size calculations, to get the size
+///    information of the inline widgets as a `List` of `PlaceholderDimensions`.
+///    Determine the positioning of the inline widgets (which is usually done by
+///    a [TextPainter] using its line break algorithm).
+///
+///  * Call [positionInlineChildren] with the positioning information of the
+///    inline widgets.
+///
+///  * Implement [RenderBox.applyPaintTransform], optionally with
+///    [defaultApplyPaintTransform].
+///
+///  * Call [paintInlineWidgets] in [RenderBox.paint] to paint the inline widgets.
+///
+///  * Call [hitTestInlineWidgets] in [RenderBox.hitTestChildren] to hit test the
+///    inline widgets.
+///
+/// See also:
+///
+///  * [WidgetSpan.extractFromInlineSpan], a helper function for extracting
+///    [WidgetSpan]s from an [InlineSpan] tree.
+mixin RenderInlineWidgetContainerDefaults on RenderBox, ContainerRenderObjectMixin<RenderBox, TextParentData> {
```

Comment:
The naming here is a bit strange, "Widget" is a concept of the layer above rendering (where this is located in), so it technically doesn't have any knowledge of what a widget is.

Same for the method on this class that contain the name widget.

Reply:
Same for the docs on this class. It should probably be rephrased to not be tied to the concept of widget (of course it can call out the WidgetSpan as an example...)